Mathlib/CategoryTheory/Comma/StructuredArrow/Final.lean

```diff
@@ -0,0 +1,103 @@
+    _ ≅ colimit <| Grothendieck.pre (functor L) R ⋙ grothendieckProj L ⋙ L ⋙ G :=
+          (Final.colimitIso (Grothendieck.pre (functor L) R) (grothendieckProj L ⋙ L ⋙ G)).symm
+    _ ≅ colimit <| Grothendieck.map (whiskerLeft _ (preFunctor L (𝟭 T))) ⋙
+          grothendieckPrecompFunctorToComma (𝟭 T) R ⋙ Comma.fst (𝟭 T) R ⋙ G :=
+            HasColimit.isoOfNatIso (NatIso.ofComponents (fun _ => Iso.refl _))
+    _ ≅ colimit <| grothendieckPrecompFunctorToComma (𝟭 T) R ⋙ Comma.fst (𝟭 T) R ⋙ G :=
+          Final.colimitIso _ _
+    _ ≅ colimit <| Grothendieck.pre (functor (𝟭 T)) R ⋙ grothendieckProj (𝟭 T) ⋙ G :=
+          HasColimit.isoOfNatIso (Iso.refl _)
+    _ ≅ colimit <| grothendieckProj (𝟭 T) ⋙ G :=
+          Final.colimitIso _ _
+    _ ≅ colimit G := (colimitIsoColimitGrothendieck (𝟭 T) G).symm
+  convert (Iso.isIso_hom i)
+  simp only [Iso.instTransIso_trans, comp_obj, grothendieckProj_obj, Grothendieck.pre_obj_base,
+    Grothendieck.pre_obj_fiber, Iso.trans_assoc, Iso.trans_hom, Iso.symm_hom, i]
+  rw [← Iso.inv_comp_eq, Iso.eq_inv_comp]
+  apply colimit.hom_ext (fun _ => by simp)
+
+end Small
+
+variable {A : Type u₁} [Category.{v₁} A] {B : Type u₁} [Category.{v₂} B]
+variable {T : Type u₁} [Category.{v₃} T]
+
+/-- A functor `L : A ⥤ T` is final if there is a final functor `R : B ⥤ T` such that for all
+`b : B`, the canonical functor `CostructuredArrow L (R.obj b) ⥤ Over (R.obj b)` is final. -/
+theorem final_of_final_CostructuredArrowtoOver (L : A ⥤ T) (R : B ⥤ T) [Final R]
+    [hB : ∀ b : B, Final (CostructuredArrow.toOver L (R.obj b))] : Final L := by
+  let sA : A ≌ AsSmall.{max u₁ u₂ u₃ v₁ v₂ v₃} A := AsSmall.equiv
+  let sB : B ≌ AsSmall.{max u₁ u₂ u₃ v₁ v₂ v₃} B := AsSmall.equiv
+  let sT : T ≌ AsSmall.{max u₁ u₂ u₃ v₁ v₂ v₃} T := AsSmall.equiv
+  let L' := sA.inverse ⋙ L ⋙ sT.functor
+  let R' := sB.inverse ⋙ R ⋙ sT.functor
+  have : ∀ (b : AsSmall B), (CostructuredArrow.toOver L' (R'.obj b)).Final := fun b => by
+    dsimp only [L', R', CostructuredArrow.toOver] at hB ⊢
+    let x := (sB.inverse ⋙ R ⋙ sT.functor).obj b
+    let F'' : CostructuredArrow (sA.inverse ⋙ L ⋙ sT.functor) x ⥤ CostructuredArrow (𝟭 _) x :=
+      map₂ (F := sA.inverse) (G := sT.inverse) (whiskerLeft (sA.inverse ⋙ L) sT.unit) (𝟙 _) ⋙
+      pre L (𝟭 T) (R.obj _) ⋙ map₂ (F := sT.functor) (G := sT.functor) (𝟙 _) (𝟙 _)
+    have : F''.Final := by
+      dsimp only [F'']
+      infer_instance
+    apply final_of_natIso (F := F'')
+    apply NatIso.ofComponents (fun X => eqToIso <| by {
+      rcases X
+      simp only [map₂, Comma.map, id_obj, comp_obj, const_obj_obj, whiskerLeft_app,
+        Discrete.natTrans_app, right_eq_id, id_map, pre, Comma.preLeft, NatTrans.id_app,
+        Category.comp_id, const_obj_map, id_eq, map_comp, Equivalence.fun_inv_map,
+        Equivalence.functor_unit_comp_assoc, Category.id_comp, comp_map, F'', x, sT]
+      congr
+      change _ ≫ eqToHom (by simp) = _
+      apply eq_of_heq
+      simp })
```

Comment:
May I interest you in the following questionable-but-perhaps-less-so-than-the-current-code code:
```suggestion
    have hsT (X) : sT.counitInv.app X = 𝟙 _ := rfl
    exact NatIso.ofComponents (fun X => CostructuredArrow.isoMk (Iso.refl _) (by simp [F'', hsT]))
```